+jrclust/+sort/@TemplateClustering/calculateTemplateSim.m

```diff
@@ -0,0 +1,14 @@
+function obj = calculateTemplateSim(obj,flagged,max_lag_s)
+    if nargin<3
+        max_lag_s=5e-4; % 0.5 ms
+        if nargin<2
+            flagged=[];
+        end
+    end
+    max_lag_samples = round(max_lag_s * obj.hCfg.sampleRate);
+    if ~isempty(flagged)
+        obj.templateSim(flagged,flagged) = calculate_simscore(obj.meanWfLocalRaw(:,:,flagged),max_lag_samples,obj.hCfg.siteNeighbors(:,obj.clusterSites(flagged)));
```

Comment:
I don't see `calculate_simscore` defined anywhere.

Reply:
This is the change that was definitely not ready for prime time. Pushed without remembering it would go in the PR. 

That said, I think it's a worthwhile addition once I clean up the code. What `calculate_simscore` does (I'll add this file to the repo) is calculate a similarity score between all clusters based on the mean waveform similarity rather than a template-based similarity score. In my experience, the template-based similarity scores often don't match what you see when you look at the actual spikes. This is for two main reasons I can identify. The first is that Kilosort has the tendency to use more than one template to account for variability in a single cell's spikes. As a result, there can be two templates with low similarity but nearly identical sets of times. From the point of view of spike sorting, I think it makes sense to treat those clusters as having a high similarity, which is better captured by calculating the similarity based on mean waveform rather than the Kilosort templates. Another reason to prefer a similarity metric based on mean waveform is that it is more robust to splitting and merging. After a cluster is split or merged, there's no obvious way to use the templates to calculate similarity anymore. Currently the code just assumes the daughter cells after a split have the same similarity scores as the parent and that a merged cluster has the max similarity with other cells as the two parents. These are obviously not good assumptions. If you recalculate the similarity scores after a split or merge using the newly computed mean waveform, you can avoid these assumptions altogether.

One option is to include code for both a template-based similarity score and a waveform-based one, and include a user-settable parameter that lets you switch between the two.

Reply:
There's a mean-waveform based similarity metric in vanilla JRCLUST, which can be toggled by pressing 'w' when the similarity plot is active -- or did you have something a bit more sophisticated in mind?

Reply:
I think what I had in mind is likely closely similar. I didn't know about this already existing feature. Let me take a look.

A related question: what are the set of possible use cases of template clustering (that is, cases where the clustering object is of class "jrclust.sort.templateClustering")? I only know of the case of a kilosort import.

Reply:
OK I looked into how JRClust calculates the mean-waveform based similarity metric. It's conceptually very similar to what I was doing, with some minor differences. On the whole, I think what JRClust was already doing was superior so I've undone all my changes related to mean-waveform based similarity metrics. I especially like how the diagonal consists of a "self-similarity" score which can be significantly below 1 when a cluster's waveform is variable. A low self-similarity score seems to be a highly reliable indicator that a cluster that needs splitting! My only suggestion would be to include calculation of a similarity score between clusters with different peak sites. Sometimes two clusters with different peak sites nonetheless need to be merged, and a high similarity score can be a useful indicator. This can be accomplished with a one-character code change, setting the default value of "peaks2" to 1 in jrclust.interfaces.Clustering.computeWaveformSim. I've implemented this change.